### DISCUSSION TITLE
Customize `SpanContext` on root span creation

### DISCUSSION THREAD

**Discussion by fmonniot** [trace4cats/trace4cats#764] [General]
Hello there,

I have this use case where my company have a legacy "logging id" (which is passed around via HTTP headers) as well as the regular B3 traces. I'd like to store this logging id in a `SpanContext` (to be able to pass it around, as well as include it in logs). I can do that relatively easily by modifying the concrete span implementation once it has been created by an `EntryPoint`. The issue I'm facing is that the original span (the one created by `Span.root` or `Span.child`, and not modified) is being sent to the `SpanCompleter`.
My idea was to create a new `EntryPoint` that would modify the `SpanContext` before creating the actual `Span`. That works well for a child span, because we have to explicitly pass the parent's context but for root spans the `SpanContext` is created automatically and cannot be modified.

All this context lead to my question: would it be possible to make the root `SpanContext` customizable? Perhaps by passing the `SpanContext` as a parameter (with [`SpanContext.root[F]`](https://github.com/trace4cats/trace4cats/blob/master/modules/kernel/src/main/scala/io/janstenpickle/trace4cats/Span.scala#L195) as the default) ?

If there is an agreement, I can open a PR with this change (or any other that are deemed preferable)

**Comment by janstenpickle** [comment]
@fmonniot I'd actually recommend using a different context that contains `Span[F]`. Using [this](https://github.com/trace4cats/trace4cats-http4s/blob/master/modules/http4s-server/src/main/scala/io/janstenpickle/trace4cats/http4s/server/Http4sResourceKleislis.scala#L24-L33=) will help you extract the logging ID from a header into any context of your choosing (that also contains `Span[F]`). In the client you can use [this](https://github.com/trace4cats/trace4cats-http4s/blob/master/modules/http4s-client/src/main/scala/io/janstenpickle/trace4cats/http4s/client/ClientSyntax.scala#L20-L27=) to convert the fields in your context to `TraceHeaders` ([see example here](https://github.com/trace4cats/trace4cats-http4s/blob/master/modules/http4s-client/src/main/scala/io/janstenpickle/trace4cats/http4s/client/ClientSyntax.scala#L18=)).

You will also need to implement a [`Provde` instance ](https://github.com/trace4cats/trace4cats/blob/v0.13.1/modules/base/src/main/scala/io/janstenpickle/trace4cats/base/context/Provide.scala)for your context so a `Trace` typeclass can be derived and so it can be injected in the http routes.

**Comment by fmonniot** [comment]
Oh, I haven't though of changing the context type. Thanks for the suggestion, it's an interesting proposition. I do have some one question and one concern about it though.

- As far as I can see, the `SpanCompleter` is invoked when the span's `Resource` is being closed. So the custom context would not be of help there, and I'd still have to put the logging id in the `Span` via an attribute. Is that correct?
- Having a custom context force the usage of MTL (-like) type classes everywhere (`Provide` & co) which I was trying to avoid as to not confuse the rest of my team (who are less comfortable with the language). I'm currently using `Trace.WithContext` which was nice in the sense that my team immediately understood what it is for.

If a custom context type is the preferred direction (and it's totally fair), I suppose that means I'll have to duplicate some of the library's feature to make them more accessible.

**Comment by janstenpickle** [comment]
> Having a custom context force the usage of MTL (-like) type classes everywhere

That doesn't have to be the case, you could can create your own typeclass to obtain your logging ID. You can either implement it directly for `Kleisli[F, YourContext[F], *]` or derive it from from `Local` [see the implementation of `Trace` for how to do this.](https://github.com/trace4cats/trace4cats/blob/master/modules/core/src/main/scala/trace4cats/Trace.scala)

> I'd still have to put the logging id in the Span via an attribute. Is that correct?

Do you want the logging ID in _every_ span? If so there are couple of ways of doing this:
1. Via the typeclass you have above; add a method called something like `addLoggingIdToSpan` which uses `accessF` to take the logging id from `YourContext[F]` (which should contain both the logging ID and Span[F], e.g. `case class YourContext[F[_]](loggingId: String, span: Span[F])`) and add it to the span attributes. The downside to this would be that you have to remember to do it inside every span you want to include the logging ID
2. Create an implementation of `Trace[F]` using `YourContext[F]` like in [`Trace`](https://github.com/trace4cats/trace4cats/blob/master/modules/core/src/main/scala/trace4cats/Trace.scala#L165-L180=) e.g. 
```scala
  implicit def localSpanInstance[F[_], G[_]](implicit
    C: Local[G, YourContext[F]],
    L: Lift[F, G],
    F: MonadCancelThrow[F],
    G: MonadCancelThrow[G]
  ): Trace.WithContext[G] = new Trace.WithContext[G] {
    def put(key: String, value: AttributeValue): G[Unit] = C.accessF(ctx => L.lift(ctx.span.put(key, value)))
    def putAll(fields: (String, AttributeValue)*): G[Unit] = C.accessF(ctx => L.lift(ctx.span.putAll(fields: _*)))
    def span[A](name: String, kind: SpanKind, errorHandler: ErrorHandler)(fa: G[A]): G[A] =
      C.accessF(ctx => ctx.span.child(name, kind, errorHandler).flatTap(span => span.put("logging-id", ctx.loggingId)).mapK(L.liftK).use(C.scope(fa)))
    def headers(toHeaders: ToHeaders): G[TraceHeaders] =
      C.access(ctx => toHeaders.fromContext(ctx.span.context))
    def setStatus(status: SpanStatus): G[Unit] = C.accessF(ctx => L.lift(ctx.span.setStatus(status)))
    def traceId: G[Option[String]] = C.access(_.span.context.traceId.show.some)
    def context: G[SpanContext] = C.access(_.span.context)
  }
```
The key part being `C.accessF(ctx => ctx.span.child(name, kind, errorHandler).flatTap(span => span.put("logging-id", ctx.loggingId)).mapK(L.liftK).use(C.scope(fa)))` on the `span` method, which adds your logging ID to every span each time a child is created.

**Comment by fmonniot** [comment]
> That doesn't have to be the case, you could can create your own type class to obtain your logging ID. 
That seems to be the direction to go, yes.

You have given me much food for thought, thank you. I'll have to take some time to ponder how to do that.

Feel free to close this issue in the meantime, I'll comment here if I have more questions. Or perhaps there is a better channel than GitHub issues, given it seems to be more of a discussion than an issue.

**Comment by janstenpickle** [comment]
> Or perhaps there is a better channel than GitHub issues, given it seems to be more of a discussion than an issue.

We've recently started using [discussions](https://github.com/trace4cats/trace4cats/discussions), you might want to continue this in the Q&A section there?

**Comment by fmonniot** [comment]
Looks like I cannot convert an issue to a discussion (https://docs.github.com/en/discussions/managing-discussions-for-your-community/moderating-discussions#converting-an-issue-to-a-discussion). Would you mind doing that ?

**Reply by janstenpickle** [reply]
@fmonniot done!
